cpp/oneapi/dal/backend/primitives/stat/cov_dpc.cpp

```diff
@@ -40,94 +39,121 @@ sycl::event correlation(sycl::queue& queue,
                 const Float m = inv_n * s * s;
-                const Float c = corr_ptr[i * p_int + i];
+                const Float c = corr_ptr[i * p + i];
 
                 means_ptr[i] = s * inv_n;
                 vars_ptr[i] = inv_n1 * (c - m);
                 tmp_ptr[i] = c;
             }
         });
     });
+}
+
+template <typename Float>
+inline sycl::event finalize_correlation(sycl::queue& q,
+                                        std::int64_t row_count,
+                                        const ndview<Float, 1>& sums,
+                                        const ndview<Float, 1>& tmp,
+                                        ndview<Float, 2>& corr,
+                                        const event_vector& deps) {
+    const std::int64_t n = row_count;
+    const std::int64_t p = sums.get_count();
+    const Float inv_n = Float(1.0 / double(n));
 
-    auto finalize_corr_event = queue.submit([&](sycl::handler& cgh) {
-        // TODO: Get information about local size from device
-        const auto range = make_multiple_nd_range_1d(p * p, 256);
-        const auto p_int = dal::detail::integral_cast<int>(p);
+    const Float* sums_ptr = sums.get_data();
+    const Float* tmp_ptr = tmp.get_mutable_data();
+    Float* corr_ptr = corr.get_mutable_data();
 
-        cgh.depends_on(mean_var_event);
+    return q.submit([&](sycl::handler& cgh) {
+        const auto range = make_multiple_nd_range_1d(detail::check_mul_overflow(p, p), //
+                                                     device_max_wg_size(q));
+
+        cgh.depends_on(deps);
         cgh.parallel_for(range, [=](sycl::nd_item<1> id) {
-            const int gi = id.get_global_id();
-            if (gi < p_int * p_int) {
-                const int i = gi / p_int;
-                const int j = gi % p_int;
+            const std::int64_t gi = id.get_global_id();
+            if (gi < p * p) {
+                const std::int64_t i = gi / p;
+                const std::int64_t j = gi % p;
                 const Float is_diag = Float(i == j);
 
                 Float c = corr_ptr[gi];
                 c -= inv_n * sums_ptr[i] * sums_ptr[j];
                 c *= sycl::rsqrt(tmp_ptr[i] * tmp_ptr[j]);
-                corr_ptr[gi] = c * (is_diag - 1) + is_diag;
+                corr_ptr[gi] = c * (is_diag - 1.0f) + is_diag;
             }
         });
     });
+}
+
+template <typename Float>
+sycl::event correlation(sycl::queue& q,
+                        const ndview<Float, 2>& data,
+                        const ndview<Float, 1>& sums,
+                        ndview<Float, 2>& corr,
+                        ndview<Float, 1>& means,
+                        ndview<Float, 1>& vars,
+                        ndview<Float, 1>& tmp,
+                        const event_vector& deps) {
+    validate_input(q, data, sums, corr, means, vars, tmp);
+
+    auto gemm_event = gemm(q, data.t(), data, corr, Float(1), Float(0), deps);
```

Comment:
Shouldn't we use `syrk`?

Reply:
For now, not. As we transfer algorithm from OpenCL to DPC++